File: Python 6-7 10-12 to/24-01/encapsulation.py
# Encapsulation -> Binding individual attributes and methods into a single unit and protecting their access from outside the class
# We have three access access modifiers (private(with double underscore), protected(with single underscore) and public)
class BankAccount:
    def __init__(self, balance):
        self.__balance = balance # private attribute
        self._acc_type = "savings" # protected attribute

    def deposit(self, amount):
        self.__balance += amount
        print(f"Amount {amount} deposited successfully!!")

    def withdraw(self, amount):
        self.__balance -= amount
        print(f"Amount {amount} withdrawn successfully!!")

    # method for checking balance
    def get_balance(self):
        return self.__balance

    
acc1 = BankAccount(500)

# private attributes access not possible
# acc1.__balance = 5000 # modify balance of acc1
# print(acc1.__balance) # check balance

acc1.deposit(1000)
print(acc1.get_balance())

print(acc1._acc_type)

File: FE 7-8 24-11 to/JS/24-01/encapsulation.js
// Encapsulation -> Binding individual data and methods into a single unit(class) and also restricting the accessing of data from outside the class
// private variables can be accessed within the class only
class BankAccount {
    #balance; // private access variable
    constructor(balance) {
        this.#balance = balance;
    }

    deposit(amount) {
        this.#balance += amount;
        console.log(`Amount ${amount} deposited..`);
    }

    // reading private data using methods
    get_balance() {
        return this.#balance;
    }
}


const acc1 = new BankAccount(1000);
// modifying the balance
// acc1.#balance = 1500; // cannot access outside the class
// console.log(acc1.#balance); // checking balance

// deposit amount
acc1.deposit(500);

// check balance
console.log(acc1.get_balance());


// Error handling using try-catch
// try block contains statements which might throw error for different inputs
// catch block contains handling statements
let a = 3;
let b = "a";
console.log(a+b);
try {
    console.log(a-c); // Error code
}
catch(err) {
    console.log(err);
}
console.log("remaining program");
